cmd/prometheus/main.go

```diff
@@ -283,7 +285,22 @@ func main() {
 		remoteStorage.ApplyConfig,
 		webHandler.ApplyConfig,
 		notifier.ApplyConfig,
-		discoveryManager.ApplyConfig,
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.ScrapeConfigs {
+				c[v.JobName] = v.ServiceDiscoveryConfig
+			}
+			return discoveryManagerScrape.ApplyConfig(c)
+		},
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
+				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config pointer as the identifier.
+				// TODO Krasi - Maybe use the slice index as the reference.
+				c[fmt.Sprintf("%p", v)] = v.ServiceDiscoveryConfig
```

Comment:
Maybe a slight overkill, but how about "<index>-<hash>"? The index would really only be there for debug-ability. The hash could be built by marshaling the `v.ServiceDiscoveryConfig` and md5 hashing it.